### PR TITLE
Create python bindings for Upload Service

### PR DESCRIPTION
**Pull request by sampierson**
This is a refactor to expose the Upload functionality as a Python module as well as the CLI.
The Python API mirrors the CLI commands: `list_areas(), select_area(), upload_file()`

### CODE REVIEW

File: hca/upload/__init__.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+def select_area(**kwargs):
+    """
+    Select a new destination for uploads, using the UUID of a formerly selected area, or the URN of a new upload area.
+
+    :param urn: An Upload Area URN.
+    :param uuid: A RFC4122-compliant ID of the upload area.
+    """
+    if 'uuid' in kwargs:
+        area = UploadArea(uuid=kwargs['uuid'])
+    elif 'urn':
```
Comment (kislyuk) [inline hca/upload/__init__.py]: `elif 'urn' in kwargs:`

File: hca/upload/__init__.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import os
+import re
+
+from .config_store import ConfigStore
+from .upload_area import UploadArea
+from .upload_area_urn import UploadAreaURN
+from .exceptions import UploadException
+from .s3_agent import S3Agent
+
+UPLOAD_BUCKET_TEMPLATE = "org-humancellatlas-upload-%s"
```
Comment (kislyuk) [inline hca/upload/__init__.py]: Can this be put in the ConfigStore, or better yet, retrieved from the upload service?

Also, can you name the parameter? (`"org-humancellatlas-upload-{stage}".format(stage="foo")`)

File: hca/upload/upload_area.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            areas = ConfigStore().areas()
+            if self.uuid not in areas:
+                raise UploadException("I'm not aware of upload area \"%s\"" % self.uuid)
+            self.urn = UploadAreaURN(areas[self.uuid])
+        elif 'urn' in kwargs:
+            self.urn = kwargs['urn']
+            self.uuid = self.urn.uuid
+            ConfigStore().add_area(self.urn)
+        else:
+            raise UploadException("you must provide a uuid or URN")
```
Comment (kislyuk) [inline hca/upload/upload_area.py]: Capitalize `You`?

File: test/upload/python_bindings/test_upload_file.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            'current_area': self.area_uuid
+        }
+        config.save()
+
+    @mock_s3
+    @reset_tweak_changes
+    def test_file_upload(self):
+        self.setup_tweak_config()
+        s3 = boto3.resource('s3')
+        s3.Bucket('org-humancellatlas-upload-test').create()
```
Comment (kislyuk) [inline test/upload/python_bindings/test_upload_file.py]: Can you parameterize the bucket name across the tests?

File: hca/upload/__init__.py (hunk 1 above)
Comment (sampierson) [reply hca/upload/__init__.py]: Good catch. Thanks.

**Comment by sampierson** [comment]
I disagree re CodeClimate.  I find it a useful tool, as we don't have any other tool measuring cognitive complexity.  In order to make code maintainable I like to keep the cognitive complexity of individual methods down.

**Comment by sampierson** [comment]
Perhaps we can turn off some of the other checks you don't like.

**Pull request merged**